packages/k2-alpine/src/components/PinInput/PinInput.tsx

```diff
@@ -0,0 +1,232 @@
+import React, { forwardRef, useImperativeHandle, useMemo, useRef } from 'react'
+import { Platform, TextInput, Vibration, ViewStyle } from 'react-native'
+import Animated, {
+  cancelAnimation,
+  runOnJS,
+  SharedValue,
+  useAnimatedStyle,
+  useSharedValue,
+  withDelay,
+  withSequence,
+  withSpring,
+  withTiming
+} from 'react-native-reanimated'
+import { TouchableOpacity } from '../Primitives'
+import { useTheme } from '../..'
+
+export const PinInput = forwardRef<PinInputActions, PinInputProps>(
+  ({ value, onChangePin, length = 6, style }, ref) => {
+    const textInputRef = useRef<TextInput>(null)
+    const wrongPinAnimation = useSharedValue(0)
+    const loadingDotAnimations = useLoadingDotAnimations(length)
+    const isLoading = useSharedValue(false)
+
+    const startLoadingAnimation = (): void => {
+      const triggerAnimations = async (): Promise<void> => {
+        const animationPromises = loadingDotAnimations.map(
+          (sharedValue, index) =>
+            new Promise<void>(resolve => {
+              sharedValue.value = withSequence(
+                withDelay(index * 50, withTiming(-15, { duration: 200 })),
+                withTiming(0, { duration: 200 })
+              )
+
+              setTimeout(() => {
+                runOnJS(resolve)()
+              }, index * 50 + 400)
+            })
+        )
+
+        return Promise.all(animationPromises).then(() => {
+          if (isLoading.value) {
+            runOnJS(triggerAnimations)()
+          }
+        })
+      }
+
+      isLoading.value = true
+      triggerAnimations()
+    }
+
+    const stopLoadingAnimation = (onComplete: () => void): void => {
+      isLoading.value = false
+      loadingDotAnimations.forEach(animation => {
+        cancelAnimation(animation)
+        animation.value = 0
+      })
+
+      runOnJS(onComplete)()
+    }
+
+    const fireWrongPinAnimation = (onComplete: () => void): void => {
+      wrongPinAnimation.value = withSequence(
+        withTiming(12, { duration: 80 }),
+        withTiming(-12, { duration: 80 }),
+        withTiming(8, { duration: 70 }),
+        withTiming(-8, { duration: 70 }),
+        withTiming(4, { duration: 60 }),
+        withTiming(-4, { duration: 60 }),
+        withSpring(0, {}, isFinished => {
+          if (isFinished) {
+            runOnJS(onComplete)()
+            runOnJS(vibratePhone)()
+          }
+        })
+      )
+    }
+
+    function vibratePhone(): void {
+      Vibration.vibrate(
+        Platform.OS === 'android'
+          ? [0, 150, 10, 150, 10, 150, 10, 150, 10, 150]
+          : [0, 10, 10, 10, 10]
```

Comment:
can you add a comment about these 2 patterns? I assume you are doing it different for each platform for a reason

Reply:
This is actually from our old code, `/app/utils/JigglyPinIndicatorHook.ts` so not sure why we use different patterns.